### PR TITLE
Add support for a legacy authority in MSAL Common

### PR DESCRIPTION
**Pull request by hectormmg**
Adds support for a legacy authority type, including authority metadata handling, cache behavior, tests, and documentation.

### CODE REVIEW

File: change/@azure-msal-browser-42470d82-f43d-4f4c-92dc-684f7f44da44.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
```
Comment (tnorling) [inline change/@azure-msal-browser-42470d82-f43d-4f4c-92dc-684f7f44da44.json]: This change file shouldn't be needed given the only changes are to markdown

File: lib/msal-common/docs/authority.md [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # Authority in MSAL
 
-In OAuth 2.0 and OpenID Connect, a client application (aka *relying party*) acquires tokens from an identity provider (IdP), such as Azure AD, to prove authentication (via ID tokens) and/or authorization (via access tokens). In this relationship, the IdP has the *authority* to issue tokens to the client application, and the client application trusts the *authority* of the IdP.
+In OAuth 2.0 and OpenID Connect, a client application (aka *relying party*) acquires tokens from an identity provider (IdP), such as Azure AD, to prove authenticity (via ID tokens) and/or authorization (via access tokens). In this relationship, the IdP has the *authority* to issue tokens to the client application, and the client application trusts the *authority* of the IdP.
```
Comment (tnorling) [inline lib/msal-common/docs/authority.md]: These aren't really synonymous - I think what was there before was correct

File: lib/msal-common/src/cache/entities/AccountEntity.ts [hunk 3]
```diff
@@ -236,12 +247,18 @@ export class AccountEntity {
...
+        logger: Logger,
+        cryptoObj: ICrypto,
+        idToken?: AuthToken
+    ): string {
 
         const accountId = idToken?.claims?.sub ? idToken.claims.sub : Constants.EMPTY_STRING;
 
         // since ADFS does not have tid and does not set client_info
-        if (authType === AuthorityType.Adfs) {
+        if (authType === AuthorityType.Adfs || authType === AuthorityType.Dsts) {
```
Comment (tnorling) [inline lib/msal-common/src/cache/entities/AccountEntity.ts:258]: Does dSTS also not have the concept of tenants?

File: lib/msal-common/src/authority/Authority.ts [hunk 4]
```diff
@@ -650,7 +681,10 @@ export class Authority {
     static replaceWithRegionalInformation(metadata: OpenIdConfigResponse, azureRegion: string): OpenIdConfigResponse {
         metadata.authorization_endpoint = Authority.buildRegionalAuthorityString(metadata.authorization_endpoint, azureRegion);
         // TODO: Enquire on whether we should leave the query string or remove it before releasing the feature
-        metadata.token_endpoint = Authority.buildRegionalAuthorityString(metadata.token_endpoint, azureRegion, "allowestsrnonmsi=true");
+        metadata.token_endpoint = Authority.buildRegionalAuthorityString(
+            metadata.token_endpoint, azureRegion,
+            "allowestsrnonmsi=true"
```
Comment (jasonnutter) [inline lib/msal-common/src/authority/Authority.ts]: Can you create a sting const for this?

File: lib/msal-common/src/cache/entities/AccountEntity.ts [hunk 5]
```diff
@@ -170,10 +170,17 @@ export class AccountEntity {
...
-             * This field should not be relied upon if a custom policy is configured to return more than 1 email.
+             * In B2C scenarios the emails claim is used instead of preferred_username and it is an array.
+             * In most cases it will contain a single email. This field should not be relied upon if a custom 
+             * policy is configured to return more than 1 email.
              */
-            account.username = idToken?.claims?.preferred_username || (idToken?.claims?.emails? idToken.claims.emails[0]: Constants.EMPTY_STRING);
+            account.username = 
+                idToken?.claims?.preferred_username ||
+                (
+                    idToken?.claims?.emails?
```
Comment (jasonnutter) [inline lib/msal-common/src/cache/entities/AccountEntity.ts]: Spacing would help readability here

**Comment by jasonnutter** [comment]
Possible to add a sample?

File: lib/msal-common/docs/authority.md [hunk 6]
```diff
@@ -93,6 +93,20 @@ const pca = new PublicClientApplication({
 
 See [ADFS](./ADFS.md) for more.
 
+### dSTS
+
+Given that dSTS also uses custom domains in its authority URLs and does not support instance discovery, dSTS authorities must also be explicitly included in the `knownAuthorities` configuration array.
+
+```javascript
+const cca = new ConfidentialClientApplication({
```
Comment (jasonnutter) [inline lib/msal-common/docs/authority.md:101]: To be clear, both public clients and confidential clients are supported?

Comment (bgavrilMS) [reply lib/msal-common/docs/authority.md:101]: Yes, I believe dSTS folks have tested CCA's client_credentials flow and PCA's integrated windows auth flow and OBO is on their roadmap.

File: lib/msal-common/src/cache/entities/AccountEntity.ts (hunk 3 above)
Comment (bgavrilMS) [reply lib/msal-common/src/cache/entities/AccountEntity.ts:258]: dSTS does have tenants, but afaik they do not support guest accounts and client_info probably isn't emitted. @neha-bhargava, do you know?

File: lib/msal-common/test/authority/Authority.spec.ts [hunk 7]
```diff
@@ -907,6 +939,20 @@ describe("Authority.ts Class Unit Tests", () => {
             expect(endpoint).toBe(`${authorityUrl}.well-known/openid-configuration`);
         });
 
+        it("DSTS authority uses v1 well-known endpoint", async () => {
+            const authorityUrl = "https://login.microsoftonline.com/dstsv2/"
```
Comment (bgavrilMS) [inline lib/msal-common/test/authority/Authority.spec.ts]: This isn't the correct authority. The authority should include the tenant. Try 


`https://login.microsoftonline.com/dstsv2/common` and 
`https://login.microsoftonline.com/dstsv2/<GUID>`

File: lib/msal-common/test/authority/Authority.spec.ts [hunk 8]
```diff
@@ -907,6 +939,20 @@ describe("Authority.ts Class Unit Tests", () => {
...
+            jest.spyOn(networkInterface, <any>"sendGetRequestAsync").mockImplementation((openIdConfigEndpoint) => {
+                // @ts-ignore
+                endpoint = openIdConfigEndpoint;
+                return DEFAULT_OPENID_CONFIG_RESPONSE;
+            });
+
+            await authority.resolveEndpointsAsync();
+            expect(endpoint).toBe(`${authorityUrl}.well-known/openid-configuration`);
+        });
+
```
Comment (bgavrilMS) [inline lib/msal-common/test/authority/Authority.spec.ts:969]: Is it possible to add a unit test based on the Fiddler trace that was shared? By mocking the HTTP response.

**Review by bgavrilMS — APPROVED** [review]
Looks good. I would add a test based on the network trace that was shared (for client_credentials).

File: lib/msal-common/test/authority/Authority.spec.ts (hunk 8 above)
Comment (hectormmg) [reply lib/msal-common/test/authority/Authority.spec.ts:969]: I'll give it a shot; I assume I'll just have to translate my manual test to a set of unit tests. Thanks for the added context and feedback!

File: lib/msal-common/src/cache/entities/AccountEntity.ts (hunk 3 above)
Comment (hectormmg) [reply lib/msal-common/src/cache/entities/AccountEntity.ts:258]: I can't remember where I read it but I'm fairly certain I did read client_info wasn't emitted (and it's not returned in the Fiddler trace either).

File: lib/msal-common/test/authority/Authority.spec.ts (hunk 7 above)
Comment (hectormmg) [reply lib/msal-common/test/authority/Authority.spec.ts]: Fixed, thanks for catching this.

File: lib/msal-common/src/authority/Authority.ts (hunk 4 above)
Comment (hectormmg) [reply lib/msal-common/src/authority/Authority.ts]: Done

File: lib/msal-common/src/cache/entities/AccountEntity.ts (hunk 5 above)
Comment (hectormmg) [reply lib/msal-common/src/cache/entities/AccountEntity.ts]: Did some refactoring for readability, too many ternaries and optional changing, it was hard to tell that the precedence went preferred name > email > empty.

**Comment by hectormmg** [comment]
> Possible to add a sample?

We don't really have a client app/tenant/account to test with, best we can do is have unit tests based on a Fiddler trace for now.

File: lib/msal-common/test/authority/Authority.spec.ts (hunk 8 above)
Comment (hectormmg) [reply lib/msal-common/test/authority/Authority.spec.ts:969]: Done!

File: change/@azure-msal-browser-42470d82-f43d-4f4c-92dc-684f7f44da44.json (hunk 1 above)
Comment (hectormmg) [reply change/@azure-msal-browser-42470d82-f43d-4f4c-92dc-684f7f44da44.json]: The check is failing for some reason, what could this be?

**Review by bgavrilMS — APPROVED** [review]

File: lib/msal-common/docs/authority.md (hunk 6 above)
Comment (jasonnutter) [reply lib/msal-common/docs/authority.md:101]: @hectormmg Can we make sure it is clear both are supported?

**Review by jasonnutter — APPROVED** [review]

**Pull request merged**